### ISSUE TITLE
Configurable formatting

### ISSUE THREAD

**Issue by thomaseizinger** [DevinR528/toml-parse#4]
Thanks for making `toml-parse`!

I'd like to use it to auto-format toml files but would like some configuration esp. in regards to whitespace.

Any plans on exposing such configurability, like not having spaces at the beginning and end of arrays for example?

As far as I can tell, it is currently not possible to configure the `Formatter` because all the fields are private. I think for my usecase it would be sufficient if I could configure the rules within the formatter.

I'll try this in a fork and report back!

**Comment by DevinR528** [comment]
Thanks! I'm open to anything that would help/allow you to use it.

One of the other problems may be that the `Formatter` only walks the `SyntaxToken`s so it makes some "lints" difficult to frame as a rule. It may be time to upgrade the `Formatter`, I started from [this](https://github.com/DevinR528/rust-analyzer/tree/space-block/crates/ra_fmt2/src) but opted for the simpler just token walking.

If all you'd need is access to the `Formatter` rules fieldf maybe adding a `new_with_rules` constructor?

**Comment by thomaseizinger** [comment]
> If all you'd need is access to the `Formatter` rules fieldf maybe adding a `new_with_rules` constructor?

From having a quick glance I don't think it is enough because all the types (e.g. `Block`) for constructing new rules are `pub(crate)`. Unless we want to make all of them public, it might be a better idea to expose some configuration and toggle those rules on and off?

**Comment by DevinR528** [comment]
Yeah, I'm fine with either! If the existing rules + turning some off/on would be enough adding some way to config would be great.

Basically I'm fine moving in any direction as long as the current functionality is maintained. I would be fine moving the rules configuration out into some other area so there isn't a hardcoded set of rules that cannot be changed. Hopefully sometime this week I'll have a chance to look into decoupling the rules from the `Formatter`.

So it sounds like you would only need to turn rules off correct? You do not need any more fine-grained rules? Basically is the set of rules currently enough?

**Comment by thomaseizinger** [comment]
What I am trying to achieve is implement [these](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/cargo.md) formatting rules in [this](https://github.com/thomaseizinger/dprint-plugin-cargo-toml) plugin so that [dprint](https://dprint.dev/) can format our `Cargo.toml` files and I have a single tool that can format all the files I care about super fast :D

[Here](https://github.com/comit-network/comit-rs/pull/3380/files?file-filters%5B%5D=.toml&file-filters%5B%5D=No+extension#diff-0eee0b873449800fa818672bb873572303eda3ceb6059a5db8f9f525800bc9bc) is the current result which already works pretty great already!

> So it sounds like you would only need to turn rules off correct? You do not need any more fine-grained rules? Basically is the set of rules currently enough?

There are more formatting rules within the Rust guide which would require more control over the rules like putting arrays on multiple lines if they exceed a certain line length. Implementing those would require more control over the rules. However, I don't plan on implementing these because IMO they defeat an important point of consistent formatting: to avoid merge conflicts.

As such what I primarily care about is the leading and trailing whitespace within arrays so if we could add a config for that, this would be sufficient for now!
